### ISSUE TITLE
number of dims don't match in permute

### ISSUE THREAD

**Issue by PonyPinkPie** [voxelmorph/voxelmorph#555]
Hi, 

my dataset is `neurite-oasis.2d.v1.0 (24.9MB)`.

**tensorflow version train.py** is ok in this dataset, but **torch version train.py** has an error in line 203 `number of dims don't match in permute` .

it seems that `d.shape` is (1, 160, 192, 1) with **4 dims**, but `permute(0, 4, 1, 2, 3)` is trying to operate 5 dims. I guess that may `d` should new a dim. However, I don't know the meaning of each dims and how to increase this one dim.

so, what should I do to solve this problem?

thank you

**Comment by adalca** [comment]
That permute operation seems to assume 3D data. You can change it to permute (0, 3, 1, 2) I think for 2D data which you are using. 

Basically keras dimensions are [batch, ..., channels] and torch is [batch, channels, ...] and I think this piece of code is just transforming between the two.

**Comment by PonyPinkPie** [comment]
Sorry, I thought it was a network issue that caused the failure to submit.

By the way, I also change `voxelmorph/troch/losses.py`,  `class grad`, `def loss(self, _, y_pred):`, 
only reserve 
`dy = torch.abs(y_pred[:, :, 1:, :] - y_pred[:, :, :-1, :])` 
and 
`dx = torch.abs(y_pred[:, :, :, 1:] - y_pred[:, :, :, :-1])`.
then train.py can run. 
Is it right?

Anyway, thank you!!!

**Comment by adalca** [comment]
yeah looks good!
